libp2p/Host.cpp

```diff
@@ -142,15 +142,17 @@ void Host::stop()
 	// called to force io_service to kill any remaining tasks it might have -
 	// such tasks may involve socket reads from Capabilities that maintain references
 	// to resources we're about to free.
+	
+	{
+		unique_lock<mutex> l(x_runTimer);
+		// ignore if already stopped/stopping, at the same time,
+		// signal run() to prepare for shutdown and reset m_timer
+		if (!m_run.exchange(false))
```

Comment:
Not sure why do you do this under `x_runTimer` lock?

Reply:
When I move this out above the block, I see
```
==================
WARNING: ThreadSanitizer: data race (pid=65004)
  Write of size 8 at 0x7ba400004df8 by main thread:
    #0 pthread_cond_destroy <null> (libtsan.so.0+0x000000029892)
    #1 dev::eth::Client::~Client() /home/yh/src/cpp-ethereum-vptr/libethereum/Client.cpp:82 (eth+0x00000022094c)
    #2 dev::eth::EthashClient::~EthashClient() /home/yh/src/cpp-ethereum-vptr/libethashseal/EthashClient.cpp:62 (eth+0x000000389795)
    #3 dev::eth::EthashClient::~EthashClient() /home/yh/src/cpp-ethereum-vptr/libethashseal/EthashClient.cpp:65 (eth+0x0000003897e8)
    #4 std::default_delete<dev::eth::Client>::operator()(dev::eth::Client*) const /usr/include/c++/7/bits/unique_ptr.h:78 (eth+0x0000004551c3)
    #5 std::unique_ptr<dev::eth::Client, std::default_delete<dev::eth::Client> >::reset(dev::eth::Client*) /usr/include/c++/7/bits/unique_ptr.h:376 (eth+0x0000004551c3)
    #6 dev::WebThreeDirect::~WebThreeDirect() /home/yh/src/cpp-ethereum-vptr/libwebthree/WebThree.cpp:91 (eth+0x0000004551c3)
    #7 main /home/yh/src/cpp-ethereum-vptr/eth/main.cpp:937 (eth+0x0000000bf8f2)

  Previous read of size 8 at 0x7ba400004df8 by thread T28:
    #0 pthread_cond_broadcast <null> (libtsan.so.0+0x00000002975f)
    #1 std::condition_variable::notify_all() <null> (libstdc++.so.6+0x0000000b5df8)
    #2 std::function<void ()>::operator()() const /usr/include/c++/7/bits/std_function.h:706 (eth+0x0000001cf4f3)
    #3 dev::eth::Signal<>::operator()() <null> (eth+0x0000001d6612)
    #4 dev::eth::TransactionQueue::manageImport_WITH_LOCK(dev::FixedHash<32u> const&, dev::eth::Transaction const&) /home/yh/src/cpp-ethereum-vptr/libethereum/TransactionQueue.cpp:171 (eth+0x00000031b951)
    #5 dev::eth::TransactionQueue::import(dev::eth::Transaction const&, dev::eth::IfDropped) /home/yh/src/cpp-ethereum-vptr/libethereum/TransactionQueue.cpp:99 (eth+0x00000031bee0)
    #6 dev::eth::TransactionQueue::verifierBody() /home/yh/src/cpp-ethereum-vptr/libethereum/TransactionQueue.cpp:396 (eth+0x00000031c49c)
    #7 operator() /home/yh/src/cpp-ethereum-vptr/libethereum/TransactionQueue.cpp:45 (eth+0x00000031ccac)
    #8 __invoke_impl<void, dev::eth::TransactionQueue::TransactionQueue(unsigned int, unsigned int)::<lambda()> > /usr/include/c++/7/bits/invoke.h:60 (eth+0x00000031cd57)
    #9 __invoke<dev::eth::TransactionQueue::TransactionQueue(unsigned int, unsigned int)::<lambda()> > /usr/include/c++/7/bits/invoke.h:95 (eth+0x00000031cd57)
    #10 _M_invoke<0> /usr/include/c++/7/thread:234 (eth+0x00000031cd57)
    #11 operator() /usr/include/c++/7/thread:243 (eth+0x00000031cd57)
    #12 _M_run /usr/include/c++/7/thread:186 (eth+0x00000031cd57)
    #13 <null> <null> (libstdc++.so.6+0x0000000bc0fe)

  As if synchronized via sleep:
    #0 nanosleep <null> (libtsan.so.0+0x00000004c5a1)
    #1 void std::this_thread::sleep_for<long, std::ratio<1l, 1000l> >(std::chrono::duration<long, std::ratio<1l, 1000l> > const&) /usr/include/c++/7/thread:376 (eth+0x0000000b723c)
    #2 stopSealingAfterXBlocks(dev::eth::Client*, unsigned int, unsigned int&) /home/yh/src/cpp-ethereum-vptr/eth/main.cpp:274 (eth+0x0000000b723c)
    #3 main /home/yh/src/cpp-ethereum-vptr/eth/main.cpp:1247 (eth+0x0000000bf614)
```

I'm confused now because `Host.cpp` doesn't appear in the stacks.

Reply:
@gumb0 forgetting the TSAN warning, I can somehow explain.  If I set `m_run` outside without the lock, the worker termination can immediately happen, `m_timerSemaphore` might be a notification, even before this thread gets the lock.  And, this thread waits on `m_timerSemaphore` forever.

Reply:
But no, `while (m_timer)` will be `false` in this case

Reply:
These stacks I think show that in `Client` condition variable `m_signalled` is being destroyed while still being used by the `m_tqReady` handler. 

Can you try moving `m_signalled` definition above `m_tqReady` in `Client.h` to change destruction order?

Reply:
Also I would name `m_timerSemaphore` something like `m_timerReset` (it's not a semaphore, it's a condition variable anyway)

Reply:
Yes, moving `m_signalled` definition worked.

Reply:
Pushing....

Reply:
Done an done.  I'm not seeing TSAN warnings.